experimental/https_attestation/src/server.rs

```diff
@@ -0,0 +1,239 @@
+    ) -> Self {
+        Self {
+            backend_uri,
+            client,
+            tee_certificate,
+            keying_material_bundle,
+            connection_status: Arc::new(Mutex::new(ConnectionStatus::NotAttested)),
+        }
+    }
+
+    async fn verify_client_assertion(&self, request: Request<Body>) -> anyhow::Result<()> {
+        let request_body = hyper::body::to_bytes(request.into_body())
+            .await
+            .context("Couldn't get request body")?
+            .to_vec();
+        let serialized_assertion =
+            String::from_utf8(request_body).context("Couldn't parse request body")?;
+        // Clients currently send empty bodies, so client assertion verification is disabled.
+        // TODO(#2048): Implement client assertion generation.
+        if !serialized_assertion.is_empty() {
+            let assertion = Assertion::from_string(&serialized_assertion)
+                .context("Couldn't deserialize client assertion")?;
+            assertion
+                .verify(&self.keying_material_bundle.client_keying_material)
+                .context("Incorrect client assertion")
+        } else {
+            Ok(())
+        }
+    }
+
+    fn generate_server_assertion(&self) -> anyhow::Result<String> {
+        let assertion = Assertion::generate(
+            &self.tee_certificate,
+            self.keying_material_bundle.server_keying_material.clone(),
+        )
+        .context("Couldn't generate server assertion")?;
+        assertion
+            .to_string()
+            .context("Couldn't serialize server assertion")
+    }
+
+    fn get_connection_status(&self) -> ConnectionStatus {
+        let connection_status = self
+            .connection_status
+            .lock()
+            .expect("Couldn't lock connection status mutex");
+        connection_status.clone()
+    }
+
+    /// Sets `Proxy::connection_status` to `ConnectionStatus::Attested`.
+    fn set_attested_connection_status(&self) {
+        let mut connection_status = self
+            .connection_status
+            .lock()
+            .expect("Couldn't lock connection status mutex");
+        *connection_status = ConnectionStatus::Attested;
+    }
+}
+
+/// Proxies HTTPS requests to an HTTP server specified by `backend_uri`.
+async fn process_request(
+    proxy: Arc<Proxy>,
+    request: Request<Body>,
+) -> Result<Response<Body>, hyper::Error> {
+    match (request.method(), request.uri().path()) {
+        (&hyper::Method::POST, "/report") => {
+            // Verify client assertion.
+            if let Err(error) = proxy.verify_client_assertion(request).await {
+                warn!("Incorrect client assertion: {:?}", error);
+                return Ok(Response::builder()
+                    .status(hyper::StatusCode::UNAUTHORIZED)
+                    .body(Body::from("Incorrect client assertion"))
+                    .unwrap());
+            }
+            proxy.set_attested_connection_status();
```

Comment:
I don't understand what this state is for; is it trying to enforce that the client calls at least once the /report endpoint? Even if it just calls it and throws away the response without reading it? I guess it's good for compatibility, but it's not really a guarantee of anything, is it?